File: autonomy/okonClient.js
'use strict';
const EventEmitter = require('events');
const net = require('net');

const PacketType = {
	get : function (val) {
		for (let [key, value] of Object.entries(this))
		  if(value === val)return key;
	},
	SET_MTR : 0xA0,
	ARM_MTR : 0xA1,
	DISARM_MTR : 0xA2,
	SET_CONTROL_MODE : 0xA3,
	SET_ACRO : 0xA4,
	SET_STABLE : 0xA5,
	SET_PID : 0xA6,
	GET_SENS : 0xB0,
	GET_DEPTH : 0xB1,
	GET_DEPTH_BYTES : 0xB2,
	GET_VIDEO_BYTES : 0xB3,
	GET_VIDEO : 0xB4,
	SET_SIM : 0xC0,
	ACK : 0xC1,
	SET_ORIEN : 0xC3,
	RST_SIM : 0xC4,
	PING : 0xC5,
	GET_CPS : 0xC6,
	HIT_NGZ : 0xC7,
	HIT_FZ : 0xC8,
	CHK_AP : 0xC9,
	ERROR : 0xCA,
	REC_STRT : 0xD0,
	REC_ST : 0xD1,
	REC_RST : 0xD2,
	GET_REC : 0xD3,
	GET_DETE : 0xDE
}

const PacketFlag = {
	get : function (val) {
		if(val === 0) return "[NONE]";
		let flags = "<";
		for (let [key, value] of Object.entries(this)){
			if(!isNaN(value)){
				//console.log(key, value, val & value);
				if(val & value)flags += " " + key;
			}
		}
		return flags + ">";
	},
	 NONE : 0,
     SERVER_ECHO : 1,
     DO_NOT_LOG_PACKET : 2,
     TEST : 128
}

function angleDifference(angle1, angle2 ) {
    let diff = Math.abs(((angle1+360)%360) - ((angle2+360)%360));
    return Math.min(diff, 360 - diff);
}

function angle0360(angle) {
    return (angle % 360 + 360) % 360;
}

function angle180(angle){
	return angle > 180 ? angle-360 : angle;
}

function angleNorm(a){
	let x = angle180(angle0360(a.x));
	let y = angle180(angle0360(a.y));
	let z = angle180(angle0360(a.z));
	if(Math.abs(x) > 90){
		y = angle180(180 + y);
		x = angle180(angle0360(180 - x));
		z = angle180(angle0360(180 + z));
	}
	return {'x':x,'y':y,'z':z};
}

class MissionControl extends EventEmitter {
	constructor(okonClient){
		super();
		this.okonClient = okonClient;
		this.time = 60000;
		this.timer = null;
		this.successes = 0;
		this.fails = 0;
	}
	
	setTimer(time){
		if(time)this.time = time;
		if(this.timer)this.resetTimer();
		this.timer = setTimeout(()=>{
			this.fails++;
			this.emit('timer');
		}, time);
	}
	
	/*end(outcome){
		if(outcome > 0) this.fails++;
		if(outcome > 0) this.fails++;
	}*/
	
	resetTimer(){
		clearTimeout(this.timer);
		this.timer = null;
	}
	
	restart(){
		
	}
}

class Okon {
	constructor(okonClient){
		this.okonClient = okonClient;
		this.sens = {
			imu: {
				rot : {x: 0, y: 0, z: 0},
				rotSpeed : {x: 0, y: 0, z: 0},
				rotAccel : {x: 0, y: 0, z: 0},
				accel : {x: 0, y: 0, z: 0}
			},
			baro : 0,
			detection : []
		};
		this.peripherals = {
			motors: { /* TODO */},
			ligths: { },
			arms: { }
		}
		this.control = {
			mode: 'unknown',
			stable: {
				targetRot: {x: 0, y: 0, z: 0},
				vel: {x: 0, y: 0, z: 0},
				targetDepth: 1.3
			},
			acro: {
				rotSpeed: {x: 0, y: 0, z: 0},
				vel: {x: 0, y: 0, z: 0}
			}
		};
		this.orien = {
			pos : {x: 0, y: 0, z: 0},
			rot : {x: 0, y: 0, z: 0}
		}
/*

- mode

  - stable
    - setRot
	- setStableVel
	- setDepth
  - acro
    - setRotSpeed
	- setAcroVel
*/
	}
	
	reachedTargetRotation(delta){
		return angleDifference(this.control.stable.targetRot.x, this.sens.imu.rot.x) < delta &&
			 angleDifference(this.control.stable.targetRot.y, this.sens.imu.rot.y) < delta &&
			 angleDifference(this.control.stable.targetRot.z, this.sens.imu.rot.z) < delta;
	}
	
	reachedTargetDepth(delta){
		return Math.abs(this.control.stable.targetDepth + this.orien.pos.y) < delta;
	}
	
	armMotors(){
		this.okonClient.send(PacketType.ARM_MTR, PacketFlag.NONE);
	}
	
	disarmMotors(){
		this.okonClient.send(PacketType.DISARM_MTR, PacketFlag.NONE);
	}
	
	reset(){
		this.control.stable = {
			targetRot: {x: 0, y: 0, z: 0},
			vel: {x: 0, y: 0, z: 0},
			targetDepth: 1
		};
		this.okonClient.send(PacketType.SET_STABLE,PacketFlag.DO_NOT_LOG_PACKET,JSON.stringify({
			rot:this.control.stable.targetRot,
			vel: this.control.stable.vel,
			depth: this.control.stable.targetDepth
		}));
	}
	
	getDetection(className){
		return this.sens.detection.filter(d => d.className === className && d.visibleInFrame);
		//for(let d of this.sens.detection) if(d.className === className && d.visibleInFrame) return d;
	}
	
	setPID(name, parameter, val, stack){
		val = Number(val);
		if(this.pids && this.pids[name] && !isNaN(this.pids[name][parameter])){
			if(stack)this.pids[name][parameter] += val;
			else this.pids[name][parameter] = val;
			this.okonClient.send(PacketType.SET_PID, PacketFlag.NONE, JSON.stringify(this.pids));
		}
	}
	
	setRot(x, y, z, add){
		if(add){
			if(!(x == null))this.control.stable.targetRot.x += x;
			if(!(y == null))this.control.stable.targetRot.y += y;
			if(!(z == null))this.control.stable.targetRot.z += z;
		}else{
			if(!(x == null))this.control.stable.targetRot.x = x;
			if(!(y == null))this.control.stable.targetRot.y = y;
			if(!(z == null))this.control.stable.targetRot.z = z;
		}
		this.control.stable.targetRot = angleNorm(this.control.stable.targetRot);
		this.okonClient.send(PacketType.SET_STABLE,PacketFlag.DO_NOT_LOG_PACKET,JSON.stringify({
			rot:this.control.stable.targetRot,
			vel: this.control.stable.vel,
			depth: this.control.stable.targetDepth
		}));
	}
	
	setRotSpeed(x, y, z, add){//TODO
		if(add){
			if(!(x == null))this.control.acro.rotSpeed.x += x;
			if(!(y == null))this.control.acro.rotSpeed.y += y;
			if(!(z == null))this.control.acro.rotSpeed.z += z;
		}else{
			if(!(x == null))this.control.acro.rotSpeed.x = x;
			if(!(y == null))this.control.acro.rotSpeed.y = y;
			if(!(z == null))this.control.acro.rotSpeed.z = z;
		}
		this.control.acro.rotSpeed = angleNorm(this.control.acro.rotSpeed);
		this.okonClient.send(PacketType.SET_ACRO,PacketFlag.DO_NOT_LOG_PACKET,JSON.stringify({
			rot:this.control.acro.rotSpeed,
			vel: this.control.acro.vel
		}));
	}
	
	setStableVel(x, y, z, add){
		if(add){
			if(!(x == null))this.control.stable.vel.x += x;
			if(!(y == null))this.control.stable.vel.y += y;
			if(!(z == null))this.control.stable.vel.z += z;
		}else{
			if(!(x == null))this.control.stable.vel.x = x;
			if(!(y == null))this.control.stable.vel.y = y;
			if(!(z == null))this.control.stable.vel.z = z;
		}
		this.okonClient.send(PacketType.SET_STABLE,PacketFlag.DO_NOT_LOG_PACKET,JSON.stringify({
			rot:this.control.stable.targetRot,
			vel: this.control.stable.vel,
			depth: this.control.stable.targetDepth
		}));
	}
	
	setAcroVel(x, y, z, add){
		if(add){
			if(!(x == null))this.control.acro.vel.x += x;
			if(!(y == null))this.control.acro.vel.y += y;
			if(!(z == null))this.control.acro.vel.z += z;
		}else{
			if(!(x == null))this.control.acro.vel.x = x;
			if(!(y == null))this.control.acro.vel.y = y;
			if(!(z == null))this.control.acro.vel.z = z;
		}
		this.okonClient.send(PacketType.SET_ACRO,PacketFlag.DO_NOT_LOG_PACKET,JSON.stringify({
			rot_speed:this.control.acro.targetRot,
			vel: this.control.acro.vel
		}));
	}
	
	setDepth(depth, add){
		if(add) this.control.stable.targetDepth += depth; 
		else this.control.stable.targetDepth = depth; 
		this.okonClient.send(PacketType.SET_STABLE,PacketFlag.DO_NOT_LOG_PACKET,JSON.stringify({
			rot:this.control.stable.targetRot,
			vel: this.control.stable.vel,
			depth: this.control.stable.targetDepth
		}));
	}
	
	setMode(mode){
		this.control.mode = mode;
		this.okonClient.send(PacketType.SET_CONTROL_MODE, PacketFlag.NONE, mode);
	}
	
	setRawMotors(/*motors object or args*/){
		//TODO GET SET packet to update or add
	}
}

class Simulation extends EventEmitter {
	constructor(okonClient){
		super();
		this.okonClient = okonClient;
		this.checkpoints = [];
	}
	
	reset(){
		this.okonClient.send(PacketType.RST_SIM, PacketFlag.NONE);
		this.okonClient.send(PacketType.DISARM_MTR, PacketFlag.NONE);
		this.okonClient.okon.reset();
	}
	
	sync(){
		this.okonClient.send(PacketType.SET_ORIEN, PacketFlag.DO_NOT_LOG_PACKET);
		this.okonClient.send(PacketType.GET_SENS, PacketFlag.DO_NOT_LOG_PACKET);
		this.okonClient.send(PacketType.SET_PID, PacketFlag.DO_NOT_LOG_PACKET);
		this.okonClient.send(PacketType.SET_CONTROL_MODE, PacketFlag.DO_NOT_LOG_PACKET);
		this.okonClient.send(PacketType.SET_STABLE, PacketFlag.DO_NOT_LOG_PACKET);
		this.okonClient.send(PacketType.GET_DETE, PacketFlag.DO_NOT_LOG_PACKET);
	}
}

class OkonClient extends EventEmitter {
	constructor(ip, port, options){ // if('customOption' in options){}
		super();
		this.options = options || {};
		
		this.simulation = new Simulation(this);
		this.okon = new Okon(this);
		this.missionControl = new MissionControl(this);
		
		this.ip = ip;
		this.port = port;
		this.socket = new net.Socket();
		this.receiveBuffer = Buffer.alloc(0,0),
		this.waitingForPacket = true;
		this.type = null;
		this.flag = null;
		this.len = null;
		this.socket.on("data", (data)=>{
			this.receiveBuffer = Buffer.concat([this.receiveBuffer, Buffer.from(data)]);
			while((this.waitingForPacket && this.receiveBuffer.length >= 6) || (!this.waitingForPacket && this.receiveBuffer.length >= this.len)){
				if(this.waitingForPacket){
					if(this.receiveBuffer.length >= 6){
						this.type = this.receiveBuffer.readUInt8(0);
						this.flag = this.receiveBuffer.readUInt8(1);
						this.len =  this.receiveBuffer.readUint32LE(2);
						this.receiveBuffer = this.receiveBuffer.slice(6);
						this.waitingForPacket = false;
					}
				} 
				if(!this.waitingForPacket){
					let jsonData = null;
					if(this.receiveBuffer.length >= this.len){
						jsonData = this.receiveBuffer.slice(0, this.len).toString('ascii');
						this.receiveBuffer = this.receiveBuffer.slice(this.len);
						this.waitingForPacket = true;
						const p = {type: this.type, flag: this.flag, json: jsonData};
						this.handleInfoPacket(p);
					}
				}
			}
		});
		this.socket.on('close', () => {
			console.log('connection closed');
			this.connected = false;
		});
	}
	
	handleInfoPacket(p){
		if(this.options.SHOW_PACKET_TIME)console.log('R', Date.now()-this.startTime,PacketType.get(p.type), p.json.length, p.json.substr(0,100));
		switch(p.type){
			case PacketType.RST_SIM:
				this.simulation.emit('reset');
				break;
			case PacketType.SET_PID:
				this.okon.pids = JSON.parse(p.json);
				break;
			case PacketType.SET_STABLE:
				let stable = JSON.parse(p.json);
				this.okon.control.stable.targetRot = stable.rot;
				this.okon.control.stable.targetDepth = stable.depth;
				this.okon.control.stable.vel = stable.vel;
				break;
			case PacketType.GET_SENS:
				let sens = JSON.parse(p.json);
				sens.rot = angleNorm(sens.rot);
				this.okon.sens.baro = sens.baro.pressure;
				this.okon.sens.imu = sens;
				break;
			case PacketType.GET_DEPTH:
				console.log('depth base64', p.json.length, '\n', p.json, '\n');
				break;
			case PacketType.GET_VIDEO:
				console.log('video base64', p.json.length, '\n', p.json, '\n');
				break;
			case PacketType.SET_ORIEN:
				let orien = JSON.parse(p.json);
				this.okon.orien.pos = orien.pos;
				this.okon.orien.rot = angleNorm(orien.rot);
				break;
			case PacketType.SET_CONTROL_MODE:
				this.okon.control.mode = p.json;
				break;
			case PacketType.ACK:
				if(p.json) console.log(JSON.parse(p.json));
				break;
			case PacketType.ERROR:
				console.log(JSON.parse(p.json));
				break;
			case PacketType.GET_DETE:
				this.okon.sens.detection = JSON.parse(p.json);
				break;
			case PacketType.GET_CPS:
				this.simulation.checkpoints = JSON.parse(p.json);
				break;
			case PacketType.HIT_NGZ:
				this.simulation.emit('hitNGZ', p.json);
				break;
			case PacketType.HIT_FZ:
				this.simulation.emit('hitFZ', p.json);
				break;
			default:
				if(!(p.flag & PacketFlag.DO_NOT_LOG_PACKET))this.emit('packet', p);
		}
	}
	
	connect(){
		this.socket.connect(this.port, this.ip, ()=>{
			this.connected = true;
			this.startTime = Date.now();
			this.simulation.sync();
			if(this.options.GET_DETE)setInterval(()=>this.send(PacketType.GET_DETE,PacketFlag.DO_NOT_LOG_PACKET), this.options.GET_DETE);
			if(this.options.GET_CPS)setInterval(()=>this.send(PacketType.GET_CPS,PacketFlag.DO_NOT_LOG_PACKET), this.options.GET_CPS);
			if(this.options.SET_ORIEN)setInterval(()=>this.send(PacketType.SET_ORIEN,PacketFlag.DO_NOT_LOG_PACKET), this.options.SET_ORIEN);
			if(this.options.GET_SENS)setInterval(()=>this.send(PacketType.GET_SENS,PacketFlag.DO_NOT_LOG_PACKET), this.options.GET_SENS);
			if(this.options.TOTAL_SYNC)setInterval(()=>this.simulation.sync(), this.options.TOTAL_SYNC);
			setTimeout(()=>this.emit('connected'), 50);
		});
	}
	
	disconnect(){
		this.connected = false;
		this.socket.end();
		this.socket.destroy();
		this.emit('disconnected');
	}
	
	send(type, flag, data){
		if(!this.connected)return;
		if(this.options.SHOW_PACKET_TIME)console.log('S', Date.now()-this.startTime,PacketType.get(type), data);
		let b = Buffer.alloc(6);
		b.writeUInt8(type, 0);
		b.writeUInt8(flag, 1);
		b.writeUInt32LE(data ? data.length : 0, 2);
		if(data)b = Buffer.concat([b, Buffer.from(data, 'ascii')]);
		this.socket.write(b);
	}
}
	
module.exports.PacketType = PacketType;
module.exports.PacketFlag = PacketFlag;
module.exports.OkonClient = OkonClient;

File: autonomy/okonClient.py
import queue
import socket
import time
from _thread import *
import threading
import json

from numpy import number


class PacketType:  # PacketType.PING
    SET_MTR = 0xA0
    ARM_MTR = 0xA1
    DISARM_MTR = 0xA2
    SET_CONTROL_MODE = 0xA3
    SET_ACRO = 0xA4
    SET_STABLE = 0xA5
    SET_PID = 0xA6
    GET_SENS = 0xB0
    GET_DEPTH = 0xB1
    GET_DEPTH_BYTES = 0xB2
    GET_VIDEO_BYTES = 0xB3
    GET_VIDEO = 0xB4
    SET_SIM = 0xC0
    ACK = 0xC1
    SET_ORIEN = 0xC3
    RST_SIM = 0xC4
    PING = 0xC5
    GET_CPS = 0xC6
    HIT_NGZ = 0xC7
    HIT_FZ = 0xC8
    CHK_AP = 0xC9
    ERROR = 0xCA
    REC_STRT = 0xD0
    REC_ST = 0xD1
    REC_RST = 0xD2
    GET_REC = 0xD3
    GET_DETE = 0xDE

    def get(val: int) -> str:  # PacketType.get(0xA5)
        for k, v in vars(PacketType).items():
            if v == val:
                return k


class PacketFlag:  # PacketType.PING
    NONE = 0
    SERVER_ECHO = 1
    DO_NOT_LOG_PACKET = 2
    TEST = 128

    def get(val: int) -> str:  # PacketFlag.get(1|2)
        flags = ""
        if val == 0:
            return "NONE "
        if val & PacketFlag.SERVER_ECHO:
            flags += "SERVER_ECHO "
        if val & PacketFlag.DO_NOT_LOG_PACKET:
            flags += "DO_NOT_LOG_PACKET "
        if val & PacketFlag.TEST:
            flags += "TEST "
        return flags


class Okon:
    def __init__(self, okon_client) -> None:
        self._okon_client = okon_client
        self.sens = {
            "imu": {
                "rot": {"x": 0, "y": 0, "z": 0},
                "rotSpeed": {"x": 0, "y": 0, "z": 0},
                "rotAccel": {"x": 0, "y": 0, "z": 0},
                "accel": {"x": 0, "y": 0, "z": 0},
            },
            "baro": 0,
            "detection": [],
        }
        self.control = {
            "mode": "unknown",
            "stable": {
                "rot": {"x": 0, "y": 0, "z": 0},
                "vel": {"x": 0, "y": 0, "z": 0},
                "depth": 1.3,
            },
            "acro": {
                "rotSpeed": {"x": 0, "y": 0, "z": 0},
                "vel": {"x": 0, "y": 0, "z": 0},
            },
            "manual": {
                "FLH": 0,
                "FLV": 0,
                "BLV": 0,
                "BLH": 0,
                "FRH": 0,
                "FRV": 0,
                "BRV": 0,
                "BRH": 0,
            },
        }
        self.orien = {"pos": {"x": 0, "y": 0, "z": 0}, "rot": {"x": 0, "y": 0, "z": 0}}
        self.pids = None  # pids get from syncing with the server

    def set_depth(self, depth: float, add: bool = False) -> None:
        old_depth = self.control["stable"]["depth"]
        if add:
            self.control["stable"]["depth"] += depth
        else:
            self.control["stable"]["depth"] = depth
        if old_depth != self.control["stable"]["depth"]:
            self._okon_client.send(
                PacketType.SET_STABLE,
                PacketFlag.DO_NOT_LOG_PACKET,
                json.dumps(self.control["stable"]),
            )

    def set_stable_vel(self, x: float = None, y: float = None, z: float = None) -> None:
        if x is not None:
            self.control["stable"]["vel"]["x"] = x
        if y is not None:
            self.control["stable"]["vel"]["y"] = y
        if z is not None:
            self.control["stable"]["vel"]["z"] = z
        self._okon_client.send(
            PacketType.SET_STABLE,
            PacketFlag.DO_NOT_LOG_PACKET,
            json.dumps(self.control["stable"]),
        )

    def arm_motors(self) -> None:
        self._okon_client.send(PacketType.ARM_MTR, PacketFlag.NONE)

    def disarm_motors(self) -> None:
        self._okon_client.send(PacketType.DISARM_MTR, PacketFlag.NONE)

    def setMode(self, mode: str) -> None:
        self.control["mode"] = mode
        self._okon_client.send(PacketType.SET_CONTROL_MODE, PacketFlag.NONE, mode)

    def reachedTargetRotation(self, delta):
        return (
            angle_difference(
                self.control["stable"]["rot"]["x"], self.sens["imu"]["rot"]["x"]
            )
            < delta
            and angle_difference(
                self.control["stable"]["rot"]["y"], self.sens["imu"]["rot"]["y"]
            )
            < delta
            and angle_difference(
                self.control["stable"]["rot"]["z"], self.sens["imu"]["rot"]["z"]
            )
            < delta
        )

    def reachedTargetDepth(self, delta):
        return (
            abs(self.control["stable"]["depth"] - self.sens["baro"] / 1000 / 9.81)
            < delta
        )

    def get_detection(self, className: str):
        return list(
            filter(
                lambda d: d["className"] == className and d["visibleInFrame"],
                self.sens["detection"],
            )
        )

    def set_stable_rot(
        self, x: float = None, y: float = None, z: float = None, add=False
    ) -> None:
        if x is not None:
            if add:
                self.control["stable"]["rot"]["x"] += x
            else:
                self.control["stable"]["rot"]["x"] = x
        if y is not None:
            if add:
                self.control["stable"]["rot"]["y"] += y
            else:
                self.control["stable"]["rot"]["y"] = y
        if z is not None:
            if add:
                self.control["stable"]["rot"]["z"] += z
            else:
                self.control["stable"]["rot"]["z"] = z
        self.control["stable"]["rot"] = angle_norm(self.control["stable"]["rot"])
        self._okon_client.send(
            PacketType.SET_STABLE,
            PacketFlag.DO_NOT_LOG_PACKET,
            json.dumps(self.control["stable"]),
        )


class Simulation:
    def __init__(self, okon_client) -> None:
        self._okon_client = okon_client
        self.checkpoints = None  # got from the server

    def reset(self) -> None:
        self._okon_client.send(PacketType.RST_SIM)

    def sync(self) -> None:
        self._okon_client.send(PacketType.SET_ORIEN, PacketFlag.DO_NOT_LOG_PACKET)
        self._okon_client.send(PacketType.GET_SENS, PacketFlag.DO_NOT_LOG_PACKET)
        self._okon_client.send(PacketType.SET_PID, PacketFlag.DO_NOT_LOG_PACKET)
        self._okon_client.send(
            PacketType.SET_CONTROL_MODE, PacketFlag.DO_NOT_LOG_PACKET
        )
        self._okon_client.send(PacketType.SET_STABLE, PacketFlag.DO_NOT_LOG_PACKET)
        self._okon_client.send(PacketType.GET_DETE, PacketFlag.DO_NOT_LOG_PACKET)


def angle_difference(angle1: float, angle2: float) -> float:
    diff = abs(((angle1 + 360) % 360) - ((angle2 + 360) % 360))
    return min(diff, 360 - diff)


def angle0360(angle: float) -> float:
    return (angle % 360 + 360) % 360


def angle180(angle: float) -> float:
    return angle - 360 if angle > 180 else angle


def angle_norm(a: dict) -> dict:
    x = angle180(angle0360(a["x"]))
    y = angle180(angle0360(a["y"]))
    z = angle180(angle0360(a["z"]))
    if abs(x) > 90:
        y = angle180(180 + y)
        x = angle180(angle0360(180 - x))
        z = angle180(angle0360(180 + z))
    return {"x": x, "y": y, "z": z}


class OkonClient:
    def __init__(self, ip, port, options=None, sync_interval=0.05, debug=True) -> None:
        self.okon = Okon(self)
        self.simulation = Simulation(self)

        self.ip = ip
        self.port = port
        self.debug = debug
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.connected = False
        self._events = dict()
        self.syncTime = time.time()
        self.sync_interval = sync_interval
        self._to_send = queue.Queue()

    def connect(self) -> bool:
        if self.debug:
            print(f"Connecting to {self.ip}:{self.port}")
        try:
            self.socket.connect((self.ip, self.port))
            self.connected = True
            start_new_thread(self._comm_thread, ())
            start_new_thread(self._sync_thread, ())
            return True
        except Exception as err:
            print(f"failed to connect {err}")
            return False

    def send(
        self, packet_type: int, packet_flag: int = PacketFlag.NONE, json: str = None
    ) -> None:
        if json == None:
            self._to_send.put((packet_type, packet_flag, 0))
        else:
            json_bytes = json.encode()
            self._to_send.put((packet_type, packet_flag, len(json_bytes), json_bytes))

    def _sync_thread(self) -> None:
        while self.connected:
            while not self._to_send.empty():
                packet = self._to_send.get()
                self._send(packet)

            if time.time() > self.syncTime + self.sync_interval:
                self.syncTime = time.time()
                self.simulation.sync()
            else:
                time.sleep(0.001)  # avg ping ~2.7ms

    def _send(self, packet: tuple) -> None:
        self.socket.sendall(packet[0].to_bytes(1, byteorder="little"))
        self.socket.sendall(packet[1].to_bytes(1, byteorder="little"))
        self.socket.sendall(packet[2].to_bytes(4, byteorder="little"))
        if packet[2] > 0:
            self.socket.sendall(packet[3])
            if self.debug and packet[1] & PacketFlag.DO_NOT_LOG_PACKET == 0:
                print(
                    f'SENT {PacketType.get(packet[0])} {PacketFlag.get(packet[1])} dat:{packet[3].decode("utf-8") }'
                )
        else:
            if self.debug and packet[1] & PacketFlag.DO_NOT_LOG_PACKET == 0:
                print(f"SENT {PacketType.get(packet[0])} {PacketFlag.get(packet[1])}")

    def _receiveAll(self, n: int) -> bytes:
        buffer = b""
        while len(buffer) != n:
            buffer += self.socket.recv(n - len(buffer))
        return buffer

    def _comm_thread(self) -> None:
        while self.connected:
            packet_typeFlag = self._receiveAll(2)
            dataLength = int.from_bytes(self._receiveAll(4), "little")
            data_bytes = self._receiveAll(dataLength)
            self._handle_packet(packet_typeFlag[0], packet_typeFlag[1], data_bytes)

    def _handle_packet(self, packet_type: int, packet_flag: int, data_bytes: bytes):
        if self.debug and packet_flag & PacketFlag.DO_NOT_LOG_PACKET == 0:
            print(
                f'RECV {PacketType.get(packet_type)} {PacketFlag.get(packet_flag)} len:{len(data_bytes)} {data_bytes.decode("utf-8")[:40]}'
            )
        if packet_type == PacketType.SET_MTR:
            manual = json.loads(data_bytes.decode())
            self.okon.control["manual"] = manual
        elif packet_type == PacketType.ARM_MTR:
            pass
        elif packet_type == PacketType.DISARM_MTR:
            pass
        elif packet_type == PacketType.SET_CONTROL_MODE:
            self.okon.control["mode"] = data_bytes.decode()
        elif packet_type == PacketType.SET_ACRO:
            acro = json.loads(data_bytes.decode())
            self.okon.control["acro"] = acro
        elif packet_type == PacketType.SET_STABLE:
            stable = json.loads(data_bytes.decode())
            self.okon.control["stable"] = stable
        elif packet_type == PacketType.SET_PID:
            pids = json.loads(data_bytes.decode())
            self.okon.pids = pids
        elif packet_type == PacketType.GET_SENS:
            sens = json.loads(data_bytes.decode())
            sens["rot "] = angle_norm(sens["rot"])
            self.okon.sens["baro"] = sens["baro"]["pressure"]
            self.okon.sens["imu"] = sens
        elif packet_type == PacketType.GET_DEPTH:
            pass
        elif packet_type == PacketType.GET_DEPTH_BYTES:
            pass
        elif packet_type == PacketType.GET_VIDEO_BYTES:
            pass
        elif packet_type == PacketType.GET_VIDEO:
            pass
        elif packet_type == PacketType.SET_SIM:
            pass
        elif packet_type == PacketType.ACK:
            pass
        elif packet_type == PacketType.SET_ORIEN:
            orien = json.loads(data_bytes.decode())
            self.okon.orien["pos"] = orien["pos"]
            self.okon.orien["rot"] = angle_norm(orien["rot"])
        elif packet_type == PacketType.RST_SIM:
            self._emit_event("simRST")
        elif packet_type == PacketType.PING:
            self._emit_event("ping", data_bytes.decode())
        elif packet_type == PacketType.GET_CPS:
            checkpoints = json.loads(data_bytes.decode())
            self.simulation.checkpoints = checkpoints
        elif packet_type == PacketType.HIT_NGZ:
            ngz = json.loads(data_bytes.decode())
            self._emit_event("hitNGZ", ngz["id"])
        elif packet_type == PacketType.HIT_FZ:
            fz = json.loads(data_bytes.decode())
            self._emit_event("hitFZ", fz["id"])
        elif packet_type == PacketType.CHK_AP:
            pass
        elif packet_type == PacketType.ERROR:
            error = json.loads(data_bytes.decode())
            self._emit_event("error", error)
        elif packet_type == PacketType.REC_STRT:
            pass
        elif packet_type == PacketType.REC_ST:
            pass
        elif packet_type == PacketType.REC_RST:
            pass
        elif packet_type == PacketType.GET_REC:
            pass
        elif packet_type == PacketType.GET_DETE:
            self.okon.sens["detection"] = json.loads(data_bytes.decode())
        else:
            self._emit_event("packet", (packet_type, packet_flag, data_bytes))

    def disconnect(self) -> None:
        self.connected = False
        self.socket.close()
        self._emit_event("disconnect")

    def on_event(self, name: str, func) -> None:
        if name in self._events:
            self._events[name].append(func)
        else:
            self._events[name] = list()
            self._events[name].append(func)

    def _emit_event(self, name: str, args=None) -> None:
        if name in self._events:
            for e in self._events[name]:
                start_new_thread(e, (args,))
